src/observable.cpp

```diff
@@ -22,35 +46,58 @@ std::istream& operator>>(std::istream& input, wxString& ws) {
 }
 
 SingletonVar* SingletonVar::getInstance(const std::string& key) {
-  static std::map<std::string, SingletonVar*> instances;
+  static std::unordered_map<std::string, SingletonVar*> instances;
 
   if (instances.find(key) == instances.end()) {
     instances[key] = new SingletonVar();
   }
   return instances[key];
 }
 
+
 /* ObservedVar implementation. */
 
-void ObservedVar::listen(wxWindow* listener, wxEventType ev_type) {
+void ObservedVar::listen(wxEvtHandler* listener, wxEventType ev_type) {
+  const auto& listeners = singleton->listeners;
+  if (listeners.find(listener) != listeners.end())
+    wxLogWarning("Duplicate window listener %",  listener);
   singleton->listeners[listener] = ev_type;
 }
 
-bool ObservedVar::unlisten(wxWindow* listener) {
+bool ObservedVar::unlisten(wxEvtHandler* listener) {
   auto& listeners = singleton->listeners;
   if (listeners.find(listener) == listeners.end()) return false;
   listeners.erase(listener);
   return true;
 }
 
-const void ObservedVar::notify() {
+const void ObservedVar::notify(const std::string& s, void* client_data) {
   auto& listeners = singleton->listeners;
   for (auto l = listeners.begin(); l != listeners.end(); l++) {
     wxCommandEvent ev(l->second);
+    ev.SetClientData(client_data);
+    ev.SetString(s);
     wxPostEvent(l->first, ev);
   }
 }
 
+const void ObservedVar::notify() { notify("", 0); }
+
+using Listener = ObservedVar::Listener;
+
+Listener ObservedVar::get_listener(wxEvtHandler* eh, wxEventType ev) {
+  return Listener(new ObservedVarListener(*this, eh, ev));
+}
+
+
+/* EventVar implementation. */
+
+std::string EventVar::autokey() {
+  static  std::atomic<unsigned long> last_ix(0);
+  return std::string("!@%/+") + std::to_string(last_ix++);
```

Comment:
Why use strings as the key at all? In all situations the string actually comes from an integer, so you could just use the integer as the key and save all the string conversions.

Reply:
Yes, could. The the from global_vars, event-vars and what not share a common "key space" so to speak. Using strings is a (tto?) simple way to handle the requirement to be unique across all keys.

Reply:
> In all situations the string actually comes from an integer,

No, not for config_var. Here the name of the variable is the key. This is also the explanation: this thing has evolved from config_var.